### PR TITLE
Just tried to make the password section better.

### PR DESCRIPTION
**Pull request by MrFlacko**
There were issues with the previous one where, if you copy and paste a large password, sometimes it doesn't display the * and displayers the text.
When I first had this come up I didn't want to continue using a stronger password, because I was afriad I would know the password after, since it would be inputted incorrectly.

This follows the traditional style of Linux passwords, when you put in the password, no text comes up when you type, and it asks you to confirm the password with no text still coming up. This should still with with the multiple variables in the script, you can use it by doing `password_input VARIABLE` then the VARIABLE would be the password that is entered.

Haven't tested it using the main script, only tested it sectioned off, since I think its a pretty small change.

### CODE REVIEW

**Comment by MrFlacko** [comment]
I'm not sure why it failed, I thought I had the latest version

**Comment by vilhelmprytz** [comment]
> I'm not sure why it failed, I thought I had the latest version

@MrFlacko Review the checks, it will tell you why it failed. It is not related to any versioning. Travis checks for code style using [shellcheck](https://github.com/koalaman/shellcheck).

I am not too sure what you are trying to achieve. I guess the `*` style password prompts could be replaced with standard "Linux-style" password prompts (i.e. not printing anything while the user types the password) but your PR also removes some functionality which I don't quite understand why.

Let me know once you've resolved the failing checks and I will be able to give it a proper review.

**Comment by MrFlacko** [comment]
Is there a way to test the code before doing a pull request with Travis
Also my goal was just to make the code a bit more efficient and nicer. I'm not sure what you mean by removing functionality except for the ability to have the different messages per password, though I think the password failed message is pretty generic.

**Comment by vilhelmprytz** [comment]
> Is there a way to test the code before doing a pull request with Travis
> Also my goal was just to make the code a bit more efficient and nicer. I'm not sure what you mean by removing functionality except for the ability to have the different messages per password, though I think the password failed message is pretty generic.

@MrFlacko You can run the tests locally, i.e. use shellcheck locally. It works with VS Code. You do not have to open another PR, you can modify this one and the tests will run again.

**Comment by MrFlacko** [comment]
Ok all done

File: install-panel.sh [hunk 1]
```diff
@@ -136,35 +136,16 @@ required_input() {
...
-          fi
-      else
-        # Add typed char to output variable.
-        result+=$char
-        # Print '*' in its stead.
-        printf '*'
-      fi
+  local result=$1
+  echo -e "Please enter the password that is used throughout the install. (Use something strong)"
+  while [[ -z "$passOriginal" || ! "$passOriginal" == "$passConfirm" ]]
```
Comment (vilhelmprytz) [inline install-panel.sh:141]: This project does not use camel case. While the rest of the project does have a few inconsistencies, I try to adhere to generally accepted naming conventions. [Google's Style Guide](https://google.github.io/styleguide/shellguide.html#s7-naming-conventions) has a few tips regarding variable naming conventions.

File: install-panel.sh [hunk 2]
```diff
@@ -136,35 +136,16 @@ required_input() {
...
+    do
+      read -rsp "Password: " passOriginal && echo
+      read -rsp "Confirm Password: " passConfirm && echo
```
Comment (vilhelmprytz) [inline install-panel.sh:144]: Why do you use `&& echo` at the end? Also, what's the thing with the `-p` prompt? From `man read` I can see that `-p` is described as.

```
Display prompt on standard error, without a trailing newline, before attempting to read any input.  The prompt is displayed only if input is coming from a terminal.
```

**Review by vilhelmprytz — CHANGES_REQUESTED** [review]
Thank you very much for your contribution! :+1: I've left some comments.

File: install-panel.sh (hunk 2 above)
Comment (MrFlacko) [reply install-panel.sh:144]: I think the -p is doesn't display anything as text but it allows text to be entered, Not sure tho
the echo is just adding a new line once you enter an input

File: install-panel.sh (hunk 1 above)
Comment (MrFlacko) [reply install-panel.sh:141]: No worries ill change this

**Comment by MrFlacko** [comment]
That should fix everything

File: install-panel.sh [hunk 3]
```diff
@@ -136,35 +136,16 @@ required_input() {
...
-        printf '*'
-      fi
+  local result=$1
+  echo -en "* ${2}\n"
+  while [[ -z "$password_main" || ! "$password_main" == "$password_confirm" ]]
+    do
+      read -rsp "Password: " password_main && echo
+      read -rsp "Confirm Password: " password_confirm && echo
+      [[ -z "$password_main" ]] && echo -e "\nError: Password cannot be empty"
+      [[ ! "$password_main" == "$password_confirm" ]] && echo -e "Error: Password do not match\n"
```
Comment (vilhelmprytz) [inline install-panel.sh:146]: These prompts do not follow the general styling of the project, i.e. using `*` before text prompts.

File: install-panel.sh [hunk 4]
```diff
@@ -136,35 +136,16 @@ required_input() {
...
-        result+=$char
-        # Print '*' in its stead.
-        printf '*'
-      fi
+  local result=$1
+  echo -en "* ${2}\n"
+  while [[ -z "$password_main" || ! "$password_main" == "$password_confirm" ]]
+    do
+      read -rsp "Password: " password_main && echo
+      read -rsp "Confirm Password: " password_confirm && echo
```
Comment (vilhelmprytz) [inline install-panel.sh:144]: You resolved it [in my previous review](https://github.com/vilhelmprytz/pterodactyl-installer/pull/120#discussion_r538926268), but I still don't understand the use of `-p`. If you look at the code as it was before, it doesn't use `-p` yet it can accept user input.

File: install-panel.sh [hunk 5]
```diff
@@ -136,35 +136,16 @@ required_input() {
...
-          fi
-      else
-        # Add typed char to output variable.
-        result+=$char
-        # Print '*' in its stead.
-        printf '*'
-      fi
+  local result=$1
+  echo -en "* ${2}\n"
+  while [[ -z "$password_main" || ! "$password_main" == "$password_confirm" ]]
```
Comment (vilhelmprytz) [inline install-panel.sh:141]: `password_main` is a rather vague variable name, can you precise it further?

File: install-panel.sh [hunk 6]
```diff
@@ -925,7 +906,7 @@ function main {
   required_input user_username "Username for the initial admin account: " "Username cannot be empty"
   required_input user_firstname "First name for the initial admin account: " "Name cannot be empty"
   required_input user_lastname "Last name for the initial admin account: " "Name cannot be empty"
-  password_input user_password "Password for the initial admin account: " "Password cannot be empty"
+  password_input user_password "Password for the initial admin account: "
```
Comment (vilhelmprytz) [inline install-panel.sh:909]: Since all the other input functions within the project have two parameters (i.e. a prompt message and a specific error message when the input is empty), I'd like the `password_input` to follow the same project styling.

**Review by vilhelmprytz — CHANGES_REQUESTED** [review]
Thanks for updating the PR. While I like the change, don't you think people would be more confused with the "empty" password prompts? Most people using this script aren't that comfortable using Linux.

File: install-panel.sh (hunk 6 above)
Comment (MrFlacko) [reply install-panel.sh:909]: Fair enough I'll change that tomorrow

File: install-panel.sh (hunk 5 above)
Comment (MrFlacko) [reply install-panel.sh:141]: do you have a suggestion?

File: install-panel.sh (hunk 4 above)
Comment (MrFlacko) [reply install-panel.sh:144]: it's -p "Confirm Password: "
so it displays the Confirm Password

File: install-panel.sh [hunk 7]
```diff
@@ -136,35 +136,16 @@ required_input() {
...
-            # Erase '*' to the left.
-            printf '\b \b' 
-          fi
-      else
-        # Add typed char to output variable.
-        result+=$char
-        # Print '*' in its stead.
-        printf '*'
-      fi
+  local result=$1
```
Comment (vilhelmprytz) [inline install-panel.sh:139]: Why did you change the name of this variable? I'd prefer the variables to be consistent with the `required_input` function :slightly_smiling_face:

File: install-panel.sh (hunk 5 above)
Comment (vilhelmprytz) [reply install-panel.sh:141]: I am not too sure. I'd suggest sticking with the same format as the `required_input` function.

**Review by vilhelmprytz — CHANGES_REQUESTED** [review]

**Comment by vilhelmprytz** [comment]
I'd love to merge this, but the requested changes are still unfixed and the PR is now very out-of-date (which means several merge conflicts). Feel free to reopen the PR at any point or submit a new one with the conflicts resolved (i.e. merging your fork with master or just creating a new branch from the current master to avoid major conflicts). Closing for now.
